src/linear_algebra/bilinear_form.lean

```diff
@@ -0,0 +1,226 @@
+instance base_ring_module : module χ χ := ring.to_module 
+instance linear_map.to_module' : module χ (β →ₗ[χ] χ) := linear_map.module
+instance linear_map.to_module₂ : module χ (β →ₗ[χ] β →ₗ[χ] χ) := linear_map.module
+
+instance to_module : module χ (bilin_form χ β) := 
+{ smul := λ c B, { bilin := λ x y, c * B.bilin x y,
+                    bilin_add_left := λ x y z, by rw [bilin_add_left, left_distrib],
+                    bilin_smul_left := λ a x y, by rw [bilin_smul_left, ←mul_assoc, mul_comm c, mul_assoc],
+                    bilin_add_right := λ x y z, by rw [bilin_add_right, left_distrib],
+                    bilin_smul_right := λ a x y, by rw [bilin_smul_right, ←mul_assoc, mul_comm c, mul_assoc],
+                    },
+  smul_add := λ c B D, by {ext, unfold bilin, rw left_distrib},
+  add_smul := λ c B D, by {ext, unfold bilin, rw right_distrib},
+  mul_smul := λ a c D, by {ext, unfold bilin, rw mul_assoc},
+  one_smul := λ B, by {ext, unfold bilin, rw one_mul},
+  zero_smul := λ B, by {ext, unfold bilin, rw zero_mul},
+  smul_zero := λ B, by {ext, unfold bilin, rw mul_zero}, 
+  }
+
+def to_linear_map : β →ₗ[χ] β →ₗ[χ] χ := 
+linear_map.mk₂ χ F.1 (bilin_add_left F) (bilin_smul_left F) (bilin_add_right F) (bilin_smul_right F)  
+
+def linear_map.to_bilin (f : β →ₗ[χ] β →ₗ[χ] χ) : bilin_form χ β := 
+{ bilin := λ x y, f x y,
+  bilin_add_left := λ x y z, eq_comm.mp (linear_map.map_add f x y) ▸ linear_map.add_apply (f x) (f y) z,
+  bilin_smul_left := λ a x y, by {rw linear_map.map_smul, rw linear_map.smul_apply, rw smul_eq_mul},
+  bilin_add_right := λ x y z, linear_map.map_add (f x) y z,
+  bilin_smul_right := λ a x y, linear_map.map_smul (f x) a y
+  }
+
+
+def bilin_linear_map_equiv : (bilin_form χ β) ≃ₗ[χ] (β →ₗ[χ] β →ₗ[χ] χ) := 
+{ to_fun := to_linear_map,
+  add := λ B D, by refl, 
+  smul := λ a B, by refl,
+  inv_fun := linear_map.to_bilin,
+  left_inv := λ B, by {ext, refl},
+  right_inv := λ B, by {ext, refl},
+}
+
+end 
+
+def is_ortho (B : bilin_form α β) (x y : β) : Prop :=
+B.bilin x y = 0
+
+lemma ortho_zero (x : β) :
+is_ortho B (0 : β) x := zero_bilin x
+
+section
+
+variables {γ : Type*} [domain γ] [module γ β] {G : bilin_form γ β}
+
+theorem ortho_mul_left {x y : β} {a : γ} (ha : a ≠ 0) :
+(is_ortho G x y) ↔ (is_ortho G (a • x) y) :=
+begin
+  dunfold is_ortho,
+  split,
+  { intros H,
+    rw [bilin_smul_left, H, ring.mul_zero] },
+
+  { intros H,
+    rw [bilin_smul_left, mul_eq_zero] at H,
+    cases H,
+    { trivial },
+
+    { exact H }}
+end
+
+theorem ortho_mul_right {x y : β} {a : γ} (ha : a ≠ 0) :
+(is_ortho G x y) ↔ (is_ortho G x (a • y)) :=
+begin
+  dunfold is_ortho,
+  split,
+  { intros H,
+    rw [bilin_smul_right, H, ring.mul_zero] },
+
+  { intros H,
+    rw [bilin_smul_right, mul_eq_zero] at H,
+    cases H,
+    { trivial },
+
+    { exact H }}
+end
+
+end
+
+end bilin_form
+
+structure reflx_bilin_form (α : Type*) (β : Type*) [ring α] [add_comm_group β] [module α β] extends bilin_form α β :=
```

Comment:
What is the reason for the `x` in `reflx`?

Reply:
removed